// pp-breadcrumbs/esm5/lib/breadcrumbs.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { Observable, BehaviorSubject, of, concat, from } from 'rxjs';
import { PpBreadcrumbsResolver } from './breadcrumbs.resolver';
import { filter, mergeMap, distinct, toArray, first } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
var PpBreadcrumbsService = /** @class */ (function () {
    function PpBreadcrumbsService(router, route, injector) {
        var _this = this;
        this.router = router;
        this.injector = injector;
        this.breadcrumbs = new BehaviorSubject([]);
        this.defaultResolver = new PpBreadcrumbsResolver();
        this.router.events.pipe(filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x instanceof NavigationEnd; }))).subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.resolveCrumbs(router.routerState.snapshot.root)
                .pipe(mergeMap((/**
             * @param {?} breadcrumbs
             * @return {?}
             */
            function (breadcrumbs) { return breadcrumbs; })), distinct((/**
             * @param {?} breadcrumb
             * @return {?}
             */
            function (breadcrumb) { return breadcrumb.text; })), toArray(), mergeMap((/**
             * @param {?} breadcrumbs
             * @return {?}
             */
            function (breadcrumbs) {
                return _this.postProcess ? _this.wrapIntoObservable(_this.postProcess(breadcrumbs)).pipe(first()) : of(breadcrumbs);
            })))
                .subscribe((/**
             * @param {?} breadcrumbs
             * @return {?}
             */
            function (breadcrumbs) {
                _this.breadcrumbs.next(breadcrumbs);
            }));
        }));
    }
    Object.defineProperty(PpBreadcrumbsService.prototype, "crumbs$", {
        get: /**
         * @return {?}
         */
        function () {
            return this.breadcrumbs;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @param {?} route
     * @return {?}
     */
    PpBreadcrumbsService.prototype.resolveCrumbs = /**
     * @private
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var crumbs$ = of([]);
        /** @type {?} */
        var data = route.routeConfig && route.routeConfig.data;
        if (data && data.breadcrumbs) {
            /** @type {?} */
            var resolver = data.breadcrumbs.prototype instanceof PpBreadcrumbsResolver ? this.injector.get(data.breadcrumbs) : this.defaultResolver;
            /** @type {?} */
            var result = resolver.resolve(route, this.router.routerState.snapshot);
            crumbs$ = this.wrapIntoObservable(result).pipe(first());
        }
        return route.firstChild ? concat(crumbs$, this.resolveCrumbs(route.firstChild)) : crumbs$;
    };
    /**
     * @private
     * @template T
     * @param {?} value
     * @return {?}
     */
    PpBreadcrumbsService.prototype.wrapIntoObservable = /**
     * @private
     * @template T
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value instanceof Observable ? value : this.isPromise(value) ? from(Promise.resolve(value)) : of((/** @type {?} */ (value)));
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    PpBreadcrumbsService.prototype.isPromise = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value && typeof value.then === 'function';
    };
    PpBreadcrumbsService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    PpBreadcrumbsService.ctorParameters = function () { return [
        { type: Router },
        { type: ActivatedRoute },
        { type: Injector }
    ]; };
    /** @nocollapse */ PpBreadcrumbsService.ngInjectableDef = i0.defineInjectable({ factory: function PpBreadcrumbsService_Factory() { return new PpBreadcrumbsService(i0.inject(i1.Router), i0.inject(i1.ActivatedRoute), i0.inject(i0.INJECTOR)); }, token: PpBreadcrumbsService, providedIn: "root" });
    return PpBreadcrumbsService;
}());
export { PpBreadcrumbsService };
if (false) {
    /** @type {?} */
    PpBreadcrumbsService.prototype.postProcess;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.breadcrumbs;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.defaultResolver;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.injector;
}
