// pp-breadcrumbs/esm5/lib/breadcrumbs.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { of } from 'rxjs';
import { Injectable } from '@angular/core';
import { template, templateSettings } from 'lodash';
templateSettings.interpolate = /{{([\s\S]+?)}}/g;
var PpBreadcrumbsResolver = /** @class */ (function () {
    function PpBreadcrumbsResolver() {
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    PpBreadcrumbsResolver.prototype.resolve = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        /** @type {?} */
        var data = route.routeConfig.data;
        /** @type {?} */
        var path = this.getFullPath(route);
        /** @type {?} */
        var rawText = typeof data.breadcrumbs === 'string' ? data.breadcrumbs : data.breadcrumbs.text || data.text || path;
        return of([
            {
                path: path,
                text: this.stringFormat(rawText, route.data)
            }
        ]);
    };
    /**
     * @param {?} route
     * @return {?}
     */
    PpBreadcrumbsResolver.prototype.getFullPath = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var relativePath = (/**
         * @param {?} segments
         * @return {?}
         */
        function (segments) { return segments.reduce((/**
         * @param {?} a
         * @param {?} v
         * @return {?}
         */
        function (a, v) { return (a += '/' + v.path); }), ''); });
        /** @type {?} */
        var fullPath = (/**
         * @param {?} routes
         * @return {?}
         */
        function (routes) { return routes.reduce((/**
         * @param {?} a
         * @param {?} v
         * @return {?}
         */
        function (a, v) { return (a += relativePath(v.url)); }), ''); });
        return fullPath(route.pathFromRoot);
    };
    /**
     * @private
     * @param {?} templateString
     * @param {?} binding
     * @return {?}
     */
    PpBreadcrumbsResolver.prototype.stringFormat = /**
     * @private
     * @param {?} templateString
     * @param {?} binding
     * @return {?}
     */
    function (templateString, binding) {
        /** @type {?} */
        var compiled = template(templateString);
        return compiled(binding);
    };
    PpBreadcrumbsResolver.decorators = [
        { type: Injectable }
    ];
    return PpBreadcrumbsResolver;
}());
export { PpBreadcrumbsResolver };
