// pp-breadcrumbs/esm2015/lib/breadcrumbs.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { of } from 'rxjs';
import { Injectable } from '@angular/core';
import { template, templateSettings } from 'lodash';
templateSettings.interpolate = /{{([\s\S]+?)}}/g;
export class PpBreadcrumbsResolver {
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    resolve(route, state) {
        /** @type {?} */
        const data = route.routeConfig.data;
        /** @type {?} */
        const path = this.getFullPath(route);
        /** @type {?} */
        const rawText = typeof data.breadcrumbs === 'string' ? data.breadcrumbs : data.breadcrumbs.text || data.text || path;
        return of([
            {
                path: path,
                text: this.stringFormat(rawText, route.data)
            }
        ]);
    }
    /**
     * @param {?} route
     * @return {?}
     */
    getFullPath(route) {
        /** @type {?} */
        const relativePath = (/**
         * @param {?} segments
         * @return {?}
         */
        (segments) => segments.reduce((/**
         * @param {?} a
         * @param {?} v
         * @return {?}
         */
        (a, v) => (a += '/' + v.path)), ''));
        /** @type {?} */
        const fullPath = (/**
         * @param {?} routes
         * @return {?}
         */
        (routes) => routes.reduce((/**
         * @param {?} a
         * @param {?} v
         * @return {?}
         */
        (a, v) => (a += relativePath(v.url))), ''));
        return fullPath(route.pathFromRoot);
    }
    /**
     * @private
     * @param {?} templateString
     * @param {?} binding
     * @return {?}
     */
    stringFormat(templateString, binding) {
        /** @type {?} */
        const compiled = template(templateString);
        return compiled(binding);
    }
}
PpBreadcrumbsResolver.decorators = [
    { type: Injectable }
];
