// pp-breadcrumbs/esm2015/lib/breadcrumbs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { PpBreadcrumbsService } from './breadcrumbs.service';
export class PpBreadcrumbsComponent {
    /**
     * @param {?} service
     */
    constructor(service) {
        this.service = service;
        this.subscriptions = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.subscriptions.push(this.service.crumbs$.subscribe((/**
         * @param {?} x
         * @return {?}
         */
        x => {
            this.crumbs = x;
        })));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscriptions.forEach((/**
         * @param {?} x
         * @return {?}
         */
        x => x.unsubscribe()));
    }
}
PpBreadcrumbsComponent.decorators = [
    { type: Component, args: [{
                selector: 'pp-breadcrumbs',
                template: "<ol *ngIf=\"crumbs.length\" class=\"breadcrumb\">\r\n  <li *ngFor=\"let crumb of crumbs; let first = first; let last = last\" [ngClass]=\"{ 'active': last }\" class=\"breadcrumb-item\">\r\n    <a *ngIf=\"!last; else lastBreadcrumb\" [routerLink]=\"crumb.path\" [innerHTML]=\"crumb.text\"></a>\r\n    <ng-template #lastBreadcrumb>\r\n      <span [innerHTML]=\"crumb.text\"></span>\r\n    </ng-template>\r\n  </li>\r\n</ol>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
PpBreadcrumbsComponent.ctorParameters = () => [
    { type: PpBreadcrumbsService }
];
if (false) {
    /** @type {?} */
    PpBreadcrumbsComponent.prototype.crumbs;
    /** @type {?} */
    PpBreadcrumbsComponent.prototype.subscriptions;
    /** @type {?} */
    PpBreadcrumbsComponent.prototype.service;
}
