// pp-breadcrumbs/esm2015/lib/breadcrumbs.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { Observable, BehaviorSubject, of, concat, from } from 'rxjs';
import { PpBreadcrumbsResolver } from './breadcrumbs.resolver';
import { filter, mergeMap, distinct, toArray, first } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class PpBreadcrumbsService {
    /**
     * @param {?} router
     * @param {?} route
     * @param {?} injector
     */
    constructor(router, route, injector) {
        this.router = router;
        this.injector = injector;
        this.breadcrumbs = new BehaviorSubject([]);
        this.defaultResolver = new PpBreadcrumbsResolver();
        this.router.events.pipe(filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x instanceof NavigationEnd))).subscribe((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            this.resolveCrumbs(router.routerState.snapshot.root)
                .pipe(mergeMap((/**
             * @param {?} breadcrumbs
             * @return {?}
             */
            breadcrumbs => breadcrumbs)), distinct((/**
             * @param {?} breadcrumb
             * @return {?}
             */
            breadcrumb => breadcrumb.text)), toArray(), mergeMap((/**
             * @param {?} breadcrumbs
             * @return {?}
             */
            breadcrumbs => {
                return this.postProcess ? this.wrapIntoObservable(this.postProcess(breadcrumbs)).pipe(first()) : of(breadcrumbs);
            })))
                .subscribe((/**
             * @param {?} breadcrumbs
             * @return {?}
             */
            breadcrumbs => {
                this.breadcrumbs.next(breadcrumbs);
            }));
        }));
    }
    /**
     * @return {?}
     */
    get crumbs$() {
        return this.breadcrumbs;
    }
    /**
     * @private
     * @param {?} route
     * @return {?}
     */
    resolveCrumbs(route) {
        /** @type {?} */
        let crumbs$ = of([]);
        /** @type {?} */
        const data = route.routeConfig && route.routeConfig.data;
        if (data && data.breadcrumbs) {
            /** @type {?} */
            const resolver = data.breadcrumbs.prototype instanceof PpBreadcrumbsResolver ? this.injector.get(data.breadcrumbs) : this.defaultResolver;
            /** @type {?} */
            const result = resolver.resolve(route, this.router.routerState.snapshot);
            crumbs$ = this.wrapIntoObservable(result).pipe(first());
        }
        return route.firstChild ? concat(crumbs$, this.resolveCrumbs(route.firstChild)) : crumbs$;
    }
    /**
     * @private
     * @template T
     * @param {?} value
     * @return {?}
     */
    wrapIntoObservable(value) {
        return value instanceof Observable ? value : this.isPromise(value) ? from(Promise.resolve(value)) : of((/** @type {?} */ (value)));
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    isPromise(value) {
        return value && typeof value.then === 'function';
    }
}
PpBreadcrumbsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PpBreadcrumbsService.ctorParameters = () => [
    { type: Router },
    { type: ActivatedRoute },
    { type: Injector }
];
/** @nocollapse */ PpBreadcrumbsService.ngInjectableDef = i0.defineInjectable({ factory: function PpBreadcrumbsService_Factory() { return new PpBreadcrumbsService(i0.inject(i1.Router), i0.inject(i1.ActivatedRoute), i0.inject(i0.INJECTOR)); }, token: PpBreadcrumbsService, providedIn: "root" });
if (false) {
    /** @type {?} */
    PpBreadcrumbsService.prototype.postProcess;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.breadcrumbs;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.defaultResolver;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PpBreadcrumbsService.prototype.injector;
}
